packages/Parallel/Application/ParallelFileProcessor.php

```diff
@@ -156,21 +175,46 @@ public function process(
             };
         };
 
-        $onExitCallableProvider = function (string $processIdentifier) use (&$systemErrors): callable {
-            return function ($exitCode, string $stdErr) use (&$systemErrors, $processIdentifier): void {
+        $onExitCallableProvider = function (string $processIdentifier) use ($useSeparateProcessPerJob, &$onDataCallableProvider, &$handleErrorCallable, &$onExitCallableProvider, &$systemErrors, &$jobIdentifiers): callable {
+            return function ($exitCode, string $stdErr) use ($processIdentifier, $useSeparateProcessPerJob, &$onDataCallableProvider, &$handleErrorCallable, &$onExitCallableProvider, &$systemErrors, &$jobIdentifiers): void {
                 $this->processPool->tryQuitProcess($processIdentifier);
 
                 if ($exitCode !== Command::SUCCESS && $exitCode !== null) {
                     $systemErrors[] = new SystemError('Child process error: ' . $stdErr);
                 }
+
+                if (! $useSeparateProcessPerJob) {
+                    return;
+                }
+
+                if ($jobIdentifiers === []) {
+                    return;
+                }
+
+                $nextProcessIdentifier = array_pop($jobIdentifiers);
+                $nextParallelProcess = $this->processPool->getProcess($nextProcessIdentifier);
+
+                $onDataCallable = $onDataCallableProvider($nextProcessIdentifier, $nextParallelProcess);
+                $onExitCallable = $onExitCallableProvider($nextProcessIdentifier);
+
+                $nextParallelProcess->start(
+                    // 1. callable on data
+                    $onDataCallable,
+
+                    // 2. callable on error
+                    $handleErrorCallable,
+
+                    // 3. callable on exit
+                    $onExitCallable
+                );
```

Comment:
when use separate process per job is enabled, here is the magic: start the next available job by its process identifier